zipline/finance/asset_restrictions.py

```diff
@@ -0,0 +1,215 @@
+    'Restriction', ['asset', 'effective_date', 'state']
+)
+
+
+RESTRICTION_STATES = enum(
+    'ALLOWED',
+    'FROZEN',
+)
+
+
+class Restrictions(with_metaclass(abc.ABCMeta)):
+    """
+    Abstract restricted list interface
+    """
+
+    @abc.abstractmethod
+    def is_restricted(self, assets, dt):
+        """
+        Is the asset restricted (RestrictionStates.FROZEN) on the given dt?
+
+        Parameters
+        ----------
+        asset : Asset of iterable of Assets
+            The asset(s) for which we are querying a restriction
+        dt : pd.Timestamp
+            The timestamp of the restriction query
+
+        Returns
+        -------
+        is_restricted : bool or pd.Series[bool] indexed by asset
+            Is the asset or assets restricted on this dt?
+
+        """
+        raise NotImplementedError('is_restricted')
+
+    def __or__(self, other_restriction):
+        """
+        Base implementation for combining two restrictions. If the right side
+        is a _UnionRestrictions, calls the overriding implementation with
+        _UnionRestrictions on the left side
+        """
+        if isinstance(other_restriction, _UnionRestrictions):
+            return _UnionRestrictions.__or__(other_restriction, self)
+        return _UnionRestrictions([self, other_restriction])
+
+
+class _UnionRestrictions(Restrictions):
+    """
+    A union of a number of sub restrictions
+
+    Parameters
+    ----------
+    sub_restrictions : iterable of Restrictions (but not _UnionRestrictions)
+        The Restrictions to be added together
+    """
+
+    def __new__(cls, sub_restrictions):
+        """
+        Returns a _UnionRestrictions defined by a list of sub_restrictions,
+        while dealing with trivial NoRestrictions cases
+        """
+        sub_restrictions = [
+            r for r in sub_restrictions if not isinstance(r, NoRestrictions)
+        ]
+        if len(sub_restrictions) == 0:
+            return NoRestrictions()
+        elif len(sub_restrictions) == 1:
+            return sub_restrictions[0]
+
+        new_instance = super(_UnionRestrictions, cls).__new__(cls)
+        new_instance.sub_restrictions = sub_restrictions
+        return new_instance
+
+    def __or__(self, other_restriction):
+        """
+        Overrides the base implementation if the left side is a
+        _UnionRestrictions. Extracts the underlying sub_restrictions from the
+        _UnionRestrictions
+        """
+        if isinstance(other_restriction, _UnionRestrictions):
+            new_sub_restrictions = \
+                self.sub_restrictions + other_restriction.sub_restrictions
+        else:
+            new_sub_restrictions = self.sub_restrictions + [other_restriction]
+
+        return _UnionRestrictions(new_sub_restrictions)
+
+    def is_restricted(self, assets, dt):
+        if isinstance(assets, Asset):
+            return any(
+                r.is_restricted(assets, dt) for r in self.sub_restrictions)
```

Comment:
Personal style note: I find it easier to read and edit expressions like this if the closing paren is on a new line by itself.